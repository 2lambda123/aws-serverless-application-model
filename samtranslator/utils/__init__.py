import copy
from typing import Any, List


def as_array(x: Any) -> List[Any]:
    """Convert value to list if it already isn't."""
    return x if isinstance(x, list) else [x]


def insert_unique(xs: Any, vs: Any):
    """
    Return list of values from `xs` extended with values from `vs` that do not
    exist in `vs`.

    Inputs are converted to lists if they already aren't.
    Does not mutate original values.
    """
    xs = as_array(copy.deepcopy(xs))
    vs = as_array(copy.deepcopy(vs))

    for v in vs:
        if v not in xs:
            xs.append(v)

    return xs
