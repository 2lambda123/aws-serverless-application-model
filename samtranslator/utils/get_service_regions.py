import boto3


def get_all_available_regions(service_name):
    """
    Wrapper function to get all available regions including ones that are not regional endpoints

    :param service_name: Name of service to get available regions like 's3' etc.
    :return:Returns a list of endpoint names (e.g., ["us-east-1"]).
    """

    return boto3._get_default_session().get_available_regions(service_name, allow_non_regional=True)


def get_available_serviceable_regions(service_name):
    """
    Wrapper function to get available regions excluding ones that are not regional endpoints

    :param service_name: Name of service to get available regions like 's3' etc.
    :return:Returns a list of endpoint names (e.g., ["us-east-1"]).
    """

    return boto3._get_default_session().get_available_regions(service_name, allow_non_regional=False)


def get_available_nonserviceable_regions(service_name):
    """
    Wrapper function to get only non serviceable regions

    :param service_name: Name of service to get available regions like 's3' etc.
    :return:Returns a list of endpoint names (e.g., ["us-east-1"]).
    """

    all_regions = boto3._get_default_session().get_available_regions(service_name, allow_non_regional=True)
    s_regions = boto3._get_default_session().get_available_regions(service_name, allow_non_regional=False)

    return list(set(all_regions) - set(s_regions))
