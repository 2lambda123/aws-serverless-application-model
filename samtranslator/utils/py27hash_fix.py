"""
"""

import ctypes
import copy
import json
import sys
import logging

from samtranslator.third_party.py27hash.hash import Hash
from six import string_types


LOG = logging.getLogger(__name__)
# Constants based on Python2.7 dictionary
# See: https://github.com/python/cpython/blob/v2.7.18/Objects/dictobject.c
MINSIZE = 8
PERTURB_SHIFT = 5

unicode_string_type = str if sys.version_info.major >= 3 else unicode


class Py27UniStr(unicode_string_type):
    """
    A string subclass to allow string be recognized as Python2 unicode string
    To preserve the instance type in string operations, we need to override certain methods
    """
    def __add__(self, other):
        return Py27UniStr(super(Py27UniStr, self).__add__(other))

    def __repr__(self):
        if sys.version_info.major >= 3:
            return "u" + super(Py27UniStr, self).__repr__()
        return super(Py27UniStr, self).__repr__()

    def upper(self):
        return Py27UniStr(super(Py27UniStr, self).upper())

    def lower(self):
        return Py27UniStr(super(Py27UniStr, self).lower())

    def replace(self, __old, __new, __count=None):
        if __count:
            return Py27UniStr(super(Py27UniStr, self).replace(__old, __new, __count))
        return Py27UniStr(super(Py27UniStr, self).replace(__old, __new))

    def split(self, sep=None, maxsplit=-1):
        return [Py27UniStr(s) for s in super(Py27UniStr, self).split(sep, maxsplit)]
