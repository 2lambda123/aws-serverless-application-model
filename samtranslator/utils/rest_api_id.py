def get_rest_api_id_string(rest_api_id):
    """
    rest_api_id can be either a string or a dictionary where the actual api id is the value at key "Ref".
    If rest_api_id is a dictionary, returns value at key "Ref". Otherwise, return rest_api_id.

    :param rest_api_id: a string or dictionary that contain the rest_api_id
    :return: string value of rest_api_id
    """
    return rest_api_id.get("Ref") if isinstance(rest_api_id, dict) else rest_api_id
