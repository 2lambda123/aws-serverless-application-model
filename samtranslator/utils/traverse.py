from typing import Any, Callable, Dict, List


class Traverse:
    """
    Base Class for traversing
    """

    def traverse_wrapper(self, input_value: Any, resolution_data: Dict[str, Any], resolver_method_name: str) -> Any:
        """
        Wraps the Driver Method which allows the resolver method to be changed as needed

        :param input_value: Any primitive type  (dict, array, string etc) whose value might contain a changed value
        :param resolution_data: Data that will help with resolution. For example, changed logical ids from hashing
        LayerVersion
        :param resolver_method_name: Name of the method that will be called to actually resolve an the function. This method
            is called with the parameters `(input, resolution_data)`.
        :return: Modified `input` with values resolved
        """
        if resolver_method_name == "DependsOn":
            return self._traverse(input_value, resolution_data, self._resolve_depends_on)

        return input_value

    # here is where I would put constant variables
    def _traverse(
        self,
        input_value: Any,
        resolution_data: Dict[str, Any],
        resolver_method: Callable[[Dict[str, Any], Any], Any],
    ) -> Any:
        """
        Driver method that performs the actual traversal of input and calls the appropriate `resolver_method` when
        to perform the resolution.

        :param input_value: Any primitive type  (dict, array, string etc) whose value might contain a changed value
        :param resolution_data: Data that will help with resolution. For example, changed logical ids from hashing
            LayerVersion
        :param resolver_method: Method that will be called to actually resolve an the function. This method
            is called with the parameters `(input, resolution_data)`.
        :return: Modified `input` with values resolved
        """

        # There is data to help with resolution. Skip the traversal altogether
        if len(resolution_data) == 0:
            return input_value

        #
        # Traversal Algorithm:
        #
        # Imagine the input dictionary/list as a tree. We are doing a Pre-Order tree traversal here where we first
        # process the root node before going to its children. Dict and Lists are the only two iterable nodes.
        # Everything else is a leaf node.
        #
        #
        # We will try to resolve if we can, otherwise return the original input. In some cases, resolving
        # an value will result in a terminal state ie. {"Depends": "foo"} could resolve to a string "Foo1234". In other
        # cases, resolving values is only partial and we might need to continue traversing the tree
        # to handle nested values in the future. All of these cases lend well towards a Pre-Order traversal where we try and
        # process the value which results in a modified sub-tree to traverse.
        #
        input_value = resolver_method(input_value, resolution_data)
        if isinstance(input_value, dict):
            return self._traverse_dict(input_value, resolution_data, resolver_method)
        if isinstance(input_value, list):
            return self._traverse_list(input_value, resolution_data, resolver_method)
        # We can iterate only over dict or list types. Primitive types are terminals

        return input_value

    def _traverse_dict(
        self,
        input_dict: Dict[str, Any],
        resolution_data: Dict[str, Any],
        resolver_method: Callable[[Dict[str, Any], Any], Any],
    ) -> Any:
        """
        Traverse a dictionary to resolves changed values on every value

        :param input_dict: Input dictionary to traverse
        :param resolution_data: Data that the `resolver_method` needs to operate
        :param resolver_method: Method that can actually resolve values, if it detects one
        :return: Modified dictionary with values resolved
        """
        for key, value in input_dict.items():
            input_dict[key] = self._traverse(value, resolution_data, resolver_method)

        return input_dict

    def _traverse_list(
        self,
        input_list: List[Any],
        resolution_data: Dict[str, Any],
        resolver_method: Callable[[Dict[str, Any], Any], Any],
    ) -> Any:
        """
        Traverse a list to resolve changed values on every element

        :param input_list: List of input
        :param resolution_data: Data that the `resolver_method` needs to operate
        :param resolver_method: Method that can actually resolves values, if it detects one
        :return: Modified list with values functions resolved
        """
        for index, value in enumerate(input_list):
            input_list[index] = self._traverse(value, resolution_data, resolver_method)

        return input_list

    def _can_handle_depends_on(self, input_dict: Dict[str, Any]) -> bool:
        """
        Checks if the input dictionary is of length one and contains "DependsOn"

        :param input_dict: the Dictionary that is attempting to be resolved
        :return boolean value of validation attempt
        """
        return isinstance(input_dict, dict) and "DependsOn" in input_dict

    def _resolve_depends_on(self, input_dict: Dict[str, Any], resolution_data: Dict[str, str]) -> Dict[str, Any]:
        """
        Resolve DependsOn when logical ids get changed when transforming (ex: AWS::Serverless::LayerVersion)

        :param input_dict: Chunk of the template that is attempting to be resolved
        :param resolution_data: Dictionary of the original and changed logical ids
        :return: Modified dictionary with values resolved
        """
        # Checks if input dict is resolvable
        if input_dict is None or not self._can_handle_depends_on(input_dict=input_dict):
            return input_dict
        # Checks if DependsOn is valid
        if not (isinstance(input_dict["DependsOn"], (list, str))):
            return input_dict
        # Check if DependsOn matches the original value of a changed_logical_id key
        for old_logical_id, changed_logical_id in resolution_data.items():
            # Done like this as there is no other way to know if this is a DependsOn vs some value named the
            # same as the old logical id. (ex LayerName is commonly the old_logical_id)
            if isinstance(input_dict["DependsOn"], list):
                for index, value in enumerate(input_dict["DependsOn"]):
                    if value == old_logical_id:
                        input_dict["DependsOn"][index] = changed_logical_id
            elif input_dict["DependsOn"] == old_logical_id:
                input_dict["DependsOn"] = changed_logical_id
        return input_dict
