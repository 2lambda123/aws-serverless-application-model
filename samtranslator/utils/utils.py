import copy
from typing import Dict, Optional, cast, Any, List


def as_array(x: Any) -> List[Any]:
    """Convert value to list if it already isn't."""
    return x if isinstance(x, list) else [x]


def insert_unique(xs: Any, vs: Any) -> List[Any]:
    """
    Return copy of `xs` extended with values of `vs` that do not exist in `xs`.

    Inputs are converted to lists if they already aren't.
    """
    xs = as_array(copy.deepcopy(xs))
    vs = as_array(copy.deepcopy(vs))

    for v in vs:
        if v not in xs:
            xs.append(v)

    return cast(List[Any], xs)  # mypy doesn't recognize it


class InvalidValueType(Exception):
    def __init__(self, relative_path: str) -> None:
        if relative_path:
            super().__init__(f"The value of '{relative_path}' should be a map")
        else:
            super().__init__("It should be a map")


def dict_deep_get(d: Any, path: str) -> Optional[Any]:
    """
    Get the value deep in the dict.

    If any value along the path doesn't exist, return None.
    If any parent node exists but is not a dict, raise InvalidValueType.
    """
    relative_path = ""
    _path_nodes = path.split(".")
    while _path_nodes:
        if d is None:
            return None
        if not isinstance(d, dict):
            raise InvalidValueType(relative_path)
        d = d.get(_path_nodes[0])
        relative_path = (relative_path + f".{_path_nodes[0]}").lstrip(".")
        _path_nodes = _path_nodes[1:]
    return d


def dict_deep_update(d: Any, dict_to_merge: Dict[str, Any]) -> None:
    """
    Update the value deep in the dict. The dict_to_merge looks like
    {
      "a.b.c": <value>
    }
    it will make d["a"]["b"]["c"] = <value>.

    If any value along the path doesn't exist, create empty dict along the way.
    If any parent node exists but is not a dict, raise InvalidValueType.
    """
    for path, value in dict_to_merge.items():
        relative_path = ""
        _path_nodes = path.split(".")
        while len(_path_nodes) > 1:
            relative_path = (relative_path + f".{_path_nodes[0]}").lstrip(".")
            d = d.setdefault(_path_nodes[0], {})
            if not isinstance(d, dict):
                raise InvalidValueType(relative_path)
            _path_nodes = _path_nodes[1:]
        d[_path_nodes[0]] = value
