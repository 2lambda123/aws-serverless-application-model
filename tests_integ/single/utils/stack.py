import json
import re

def verify_stack_resources(expected_file_path, stack_resources):
    with open(expected_file_path, 'r') as expected_data:
        expected_resources = _sort_resources(json.load(expected_data))
    parsed_resources = _sort_resources(stack_resources['StackResourceSummaries'])

    if len(expected_resources) != len(parsed_resources):
        return False

    for i, in enumerate(expected_resources):
        exp = expected_resources[i]
        parsed = parsed_resources[i]
        if parsed["ResourceStatus"] != "CREATE_COMPLETE":
            return False
        if re.fullmatch(exp["LogicalResourceId"], parsed["LogicalResourceId"]) is None:
            return False
        if exp["ResourceType"] != parsed["ResourceType"]:
            return False
    return True

def _sort_resources(resources):
    return sorted(resources, key=lambda d: d["LogicalResourceId"])
