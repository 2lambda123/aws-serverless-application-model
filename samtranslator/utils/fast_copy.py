import copy
import json
from os import stat

from samtranslator.utils.py27hash_fix import Py27Dict, Py27Keys, Py27UniStr

_PY2_MARKER = "\xef\xbb\xbf"
_KEYORDER_DUMMY_MARKER = _PY2_MARKER + "DUMMY"


def encode_py27(data):
    if isinstance(data, Py27UniStr):
        return _PY2_MARKER + data
    if isinstance(data, Py27Dict):
        result = {encode_py27(key): encode_py27(value) for key, value in data.items()}
        keylist = data.keylist
        result[_PY2_MARKER] = {
            "keylist": {
                "debug": keylist.debug,
                "size": keylist.size,
                "fill": keylist.fill,
                "mask": keylist.mask,
                "keyorder": encode_keyorder(keylist.keyorder),
            }
        }
        return result

    if isinstance(data, list):
        return [encode_py27(item) for item in data]
    if isinstance(data, dict):
        return {encode_py27(key): encode_py27(value) for key, value in data.items()}
    return data


def encode_keyorder(keyorder):
    if keyorder is None:
        return None
    if keyorder == {}:
        return {}

    encoded_keyorder = dict()
    if keyorder:
        for i in keyorder:
            if keyorder[i] == Py27Keys.DUMMY:
                encoded_keyorder[i] = _KEYORDER_DUMMY_MARKER
    return encoded_keyorder


def decode_keyorder(keyorder, keys):
    if keyorder is None:
        return None
    if keyorder == {}:
        return {}

    decoded_keyorder = dict()
    for i in keyorder:
        if keyorder[i] == _KEYORDER_DUMMY_MARKER:
            decoded_keyorder[int(i)] = Py27Keys.DUMMY
        else:
            decoded_keyorder[int(i)] = keys[keys.index(keyorder[i])]
    return decoded_keyorder


def decode_py27(data):
    if isinstance(data, str) and data.startswith(_PY2_MARKER):
        return Py27UniStr(data[3:])
    if isinstance(data, list):
        return [decode_py27(item) for item in data]
    if isinstance(data, dict):
        if _PY2_MARKER in data:
            py27Dict = Py27Dict(
                {decode_py27(key): decode_py27(value) for key, value in data.items() if key != _PY2_MARKER}
            )
            state = data[_PY2_MARKER]
            keylist_state = state["keylist"]
            del data[_PY2_MARKER]

            keylist = Py27Keys()
            keylist.debug = keylist_state["debug"]
            keylist.size = keylist_state["size"]
            keylist.fill = keylist_state["fill"]
            keylist.mask = keylist_state["mask"]
            keylist.keyorder = decode_keyorder(keylist_state["keyorder"], py27Dict.keys())
            keylist = keylist.copy()

            py27Dict.keylist = keylist

            return py27Dict
        return {decode_py27(key): decode_py27(value) for key, value in data.items()}
    return data


def fast_copy(data):
    encoded_string = json.dumps(encode_py27(data))
    decoded_data = decode_py27(json.loads(encoded_string))
    decoded_data = copy.deepcopy(data)
    return decoded_data
