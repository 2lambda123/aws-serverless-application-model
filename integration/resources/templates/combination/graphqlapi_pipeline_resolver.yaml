Transform: AWS::Serverless-2016-10-31
Resources:
  MyApiKey:
    Type: AWS::AppSync::ApiKey
    Properties:
      ApiId: !GetAtt SuperCoolAPI.ApiId

  DynamoDBPostsTable:
    Type: AWS::DynamoDB::Table
    Properties:
      TableName: posts-table
      AttributeDefinitions:
      - AttributeName: id
        AttributeType: S
      KeySchema:
      - AttributeName: id
        KeyType: HASH
      ProvisionedThroughput:
        ReadCapacityUnits: 5
        WriteCapacityUnits: 5

  DynamoDBPostsLogTable:
    Type: AWS::DynamoDB::Table
    Properties:
      TableName: posts-log-table
      AttributeDefinitions:
      - AttributeName: id
        AttributeType: S
      KeySchema:
      - AttributeName: id
        KeyType: HASH
      ProvisionedThroughput:
        ReadCapacityUnits: 5
        WriteCapacityUnits: 5

  SuperCoolAPI:
    Type: AWS::Serverless::GraphQLApi
    Properties:
      Name: PradsApi
      SchemaInline: |
        schema {
            query: Query
            mutation: Mutation
        }

        type Query {
            getPost(id: String!): Post
        }

        type Mutation {
            addPost(
            author: String!,
            title: String!,
            content: String!,
            url: String!
            ): Post!
        }

        type Post {
            id: String!
            author: String
            title: String
            content: String
            url: String
            ups: Int!
            downs: Int!
            version: Int!
        }
      Auth:
        Type: API_KEY
      DynamoDBDataSources:
        PostsDataSource:
          TableName: !Ref DynamoDBPostsTable
          TableArn: !GetAtt DynamoDBPostsTable.Arn
        PostsLogDataSource:
          TableName: !Ref DynamoDBPostsLogTable
          TableArn: !GetAtt DynamoDBPostsLogTable.Arn
      AppSyncResolvers:
        Mutation:
          addPost:
            Runtime:
              Name: APPSYNC_JS
              Version: 1.0.0
            Functions:
            - createPostLogInTable:
                DataSource: PostsLogDataSource
                InlineCode: |
                  import { util } from '@aws-appsync/utils';

                  export function request(ctx) {
                    const id = util.autoId();
                    const author = ctx.args.author
                    const time = "1234567890"
                    return dynamodbPutRequest({ key: {id}, values: {"author": author, "time": time}});
                  }

                  export function response(ctx) {
                    return ctx.result;
                  }

                  /**
                  * Helper function to create a new item
                  * @returns a PutItem request
                  */
                  function dynamodbPutRequest({key, values}) {
                    return {
                      operation: 'PutItem',
                      key: util.dynamodb.toMapValues(key),
                      attributeValues: util.dynamodb.toMapValues(values),
                    };
                  }
            - createPostInTable:
                DataSource: PostsDataSource
                InlineCode: |
                  import { util } from '@aws-appsync/utils';

                  export function request(ctx) {
                    const id = util.autoId();
                    const { ...values } = ctx.args;
                    values.ups = 1;
                    values.downs = 0;
                    values.version = 1;
                    return dynamodbPutRequest({ key: {id}, values });
                  }

                  export function response(ctx) {
                    return ctx.result;
                  }

                  /**
                  * Helper function to create a new item
                  * @returns a PutItem request
                  */
                  function dynamodbPutRequest({key, values}) {
                    return {
                      operation: 'PutItem',
                      key: util.dynamodb.toMapValues(key),
                      attributeValues: util.dynamodb.toMapValues(values),
                    };
                  }                
        Query:
          getPost:
            DataSource: PostsDataSource
            Runtime:
              Name: APPSYNC_JS
              Version: 1.0.0
            Functions:
            - getPostFromTable:
                InlineCode: |
                  import { util } from '@aws-appsync/utils';

                  export function request(ctx) {
                    return dynamoDBGetItemRequest({ id: ctx.args.id });
                  }

                  export function response(ctx) {
                    return ctx.result;
                  }

                  /**
                  * A helper function to get a DynamoDB item
                  */
                  function dynamoDBGetItemRequest(key) {
                    return {
                      operation: 'GetItem',
                      key: util.dynamodb.toMapValues(key),
                    };
                  }
Outputs:
  SuperCoolAPI:
    Description: AppSync API
    Value: !GetAtt SuperCoolAPI.GraphQLUrl
  MyApiKey:
    Description: API Key for authentication
    Value: !GetAtt MyApiKey.ApiKey
Metadata:
  SamTransformTest: true
