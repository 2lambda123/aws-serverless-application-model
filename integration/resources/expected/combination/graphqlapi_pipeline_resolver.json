[
  {
    "LogicalResourceId": "DynamoDBPostsTable",
    "ResourceType": "AWS::DynamoDB::Table"
  },
  {
    "LogicalResourceId": "DynamoDBPostsLogTable",
    "ResourceType": "AWS::DynamoDB::Table"
  },
  {
    "LogicalResourceId": "MyApiKey",
    "ResourceType": "AWS::AppSync::ApiKey"
  },
  {
    "LogicalResourceId": "SuperCoolAPI",
    "ResourceType": "AWS::AppSync::GraphQLApi"
  },
  {
    "LogicalResourceId": "SuperCoolAPISchema",
    "ResourceType": "AWS::AppSync::GraphQLSchema"
  },
  {
    "LogicalResourceId": "SuperCoolAPICloudWatchRole",
    "ResourceType": "AWS::IAM::Role"
  },
  {
    "LogicalResourceId": "SuperCoolAPIDynamoDBDataSourcePostsDataSource",
    "ResourceType": "AWS::AppSync::DataSource"
  },
  {
    "LogicalResourceId": "SuperCoolAPIDynamoDBDataSourcePostsDataSourceRole",
    "ResourceType": "AWS::IAM::Role"
  },
  {
    "LogicalResourceId": "SuperCoolAPIDynamoDBDataSourcePostsDataSourceToTableConnectorPolicy",
    "ResourceType": "AWS::IAM::ManagedPolicy"
  },
  {
    "LogicalResourceId": "SuperCoolAPIDynamoDBDataSourcePostsLogDataSource",
    "ResourceType": "AWS::AppSync::DataSource"
  },
  {
    "LogicalResourceId": "SuperCoolAPIDynamoDBDataSourcePostsLogDataSourceRole",
    "ResourceType": "AWS::IAM::Role"
  },
  {
    "LogicalResourceId": "SuperCoolAPIDynamoDBDataSourcePostsLogDataSourceToTableConnectorPolicy",
    "ResourceType": "AWS::IAM::ManagedPolicy"
  },
  {
    "LogicalResourceId": "SuperCoolAPIcreatePostLogInTable",
    "ResourceType": "AWS::AppSync::FunctionConfiguration"
  },
  {
    "LogicalResourceId": "SuperCoolAPIcreatePostInTable",
    "ResourceType": "AWS::AppSync::FunctionConfiguration"
  },
  {
    "LogicalResourceId": "SuperCoolAPIaddPost",
    "ResourceType": "AWS::AppSync::Resolver"
  },
  {
    "LogicalResourceId": "SuperCoolAPIgetPostFromTable",
    "ResourceType": "AWS::AppSync::FunctionConfiguration"
  },
  {
    "LogicalResourceId": "SuperCoolAPIgetPost",
    "ResourceType": "AWS::AppSync::Resolver"
  }
]
